scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala

```diff
@@ -1,87 +1,167 @@
+    if (close.pos.start.line != bodyEnd.end.line) {
+      import ctx.tokenTraverser._
+      val next = nextToken(close)
+      if (next.is[LF])
+        builder += TokenPatch.Remove(next)
+    }
+
+  private def processBlock(
+      b: Term.Block)(implicit builder: PatchBuilder, ctx: RewriteCtx): Unit =
+    if (b.tokens.nonEmpty) {
+      val open = b.tokens.head
+      if (open.is[LeftBrace]) {
+        val close = b.tokens.last
+        if (removeBlock(b) && close.is[RightBrace]) {
+          val endPos = if (b.stats.isEmpty) b.pos else b.stats.last.pos
+          removeTrailingLF(endPos, close)
+          builder += TokenPatch.Remove(open)
+          builder += TokenPatch.Remove(close)
         }
-        val next = nextToken(close)
-        if (next.is[LF] &&
-          close.pos.start.line != bodyStatement.pos.end.line)
-          builder += TokenPatch.Remove(next)
+      }
+    }
+
+  private def removeBlock(b: Term.Block)(implicit ctx: RewriteCtx): Boolean = {
+    def exactlyOneStatement = b.stats.lengthCompare(1) == 0
+    b.parent.exists {
+
+      case _: Case =>
+        settings.generalExpressions
+
+      case _: Term.Apply =>
+        // Example: as.map { _.toString }
+        // Leave this alone for now.
+        // In future there should be an option to surround such expressions with parens instead of braces
+        false
 
-        builder += TokenPatch.Remove(open)
-        builder += TokenPatch.Remove(close)
+      case d: Defn.Def =>
+        def disqualifiedByUnit =
+          !settings.includeUnitMethods && d.decltpe.exists(_.syntax == "Unit")
+        def innerOk =
+          b.stats.head match {
+            case _: Term.Function | _: Defn => false
+            case _ => true
+          }
+        settings.methodBodies &&
+        exactlyOneStatement &&
+        blockSizeIsOk(b) &&
+        innerOk &&
+        !isProcedureSyntax(d) &&
+        !disqualifiedByUnit
+
+      case _ =>
+        settings.generalExpressions &&
+          exactlyOneStatement &&
+          blockSizeIsOk(b) &&
+          !retainSingleStatBlock(b)
     }
-    builder.result()
   }
+
+  /** Some blocks look redundant but aren't */
+  private def retainSingleStatBlock(b: Term.Block): Boolean =
+    b.parent.exists {
+      case parentIf: Term.If =>
+        // if (a) { if (b) c } else d
+        //   ↑ cannot be replaced by ↓
+        // if (a) if (b) c else d
+        //   which would be equivalent to
+        // if (a) { if (b) c else d }
+        def insideIfThen = parentIf.thenp eq b
+        def parentIfHasAnElse = parentIf.elsep.tokens.nonEmpty
+        def blockIsIfWithoutElse = b.stats.head match {
+          case childIf: Term.If => childIf.elsep.tokens.isEmpty
+          case _ => false
+        }
+        insideIfThen && parentIfHasAnElse && blockIsIfWithoutElse
+
+      case parent =>
+        val side = parent match {
+          case t: Term.ApplyInfix
+              if t.args.lengthCompare(1) == 0 && (t.args.head eq b) =>
```

Comment:
Note @olafurpg , I've changed this from `case Term.ApplyInfix(_, _, _, arg :: Nil) if arg eq b =>` to

1) be a little more runtime-efficient by bypassing `unappy`s
2) this now works as expected if args isn't a `List`